router/io-spec.gobra

```diff
@@ -543,3 +543,48 @@ pure func uInfo(hopfields seq[io.IO_HF], currHFIdx int, consDir bool) set[io.IO_
 						hvfSet(hopfields[currHFIdx]) :
 						hvfSet(hopfields[currHFIdx-1])))
 }
+
+ghost
+decreases
+pure func ifsToIO_ifs(ifs uint16) option[io.IO_ifs]{
+    return ifs == 0 ? none[io.IO_ifs] : some(io.IO_ifs(ifs))
+}
+
+ghost 
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
+decreases
+pure func validPktMetaHdr(raw []byte) bool {
+	return len(raw) > 4 && 
+		unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _) in 
+		let hdr := binary.BigEndian.Uint32(raw[0:4]) in
+		let metaHdr := scion.DecodedFrom(hdr) in
+		let seg1 := int(metaHdr.SegLen[0]) in
+		let seg2 := int(metaHdr.SegLen[1]) in
+		let seg3 := int(metaHdr.SegLen[2]) in
+		let base := scion.Base{metaHdr, 
+			numInfoFields(seg1, seg2, seg3),
+			seg1 + seg2 + seg3} in
+		metaHdr.InBounds() &&
+		0 < metaHdr.SegLen[0] &&
+		base.ValidCurrInfSpec() &&
+		base.ValidCurrHfSpec() &&
+		len(raw) >= pktLen(seg1, seg2, seg3)
+}
+
+ghost 
+requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
+ensures val.isIO_val_Unsupported
+ensures val.IO_val_Unsupported_1 == ifsToIO_ifs(ingressID)
+decreases
+pure func absIO_val_Unsupported(raw []byte, ingressID uint16) (ghost val io.IO_val)
```

Comment:
I don't think so. It is not really clear how an unsupported packet would look like. We don't prove any properties for unsupported packets and any array of bytes can always be converted to a unsupported packet.

Reply:
Hmm, I guess that we would need to clarify the purpose of the second parameter in the following constructor:
```
	// Unsupported
	IO_val_Unsupported {
		IO_val_Unsupported_1 option[IO_ifs]
		IO_val_Unsupported_2 TypeParameter // <-- (TODO) yet to replace by the concrete type used
	}
```
I think it would be better to do it now than to find it later. Out of curiosity, why did you decide with this approach instead of return an option?

Reply:
If we want to use `option` we would have to change the io-spec significantly. We are limited by the return value of `dp3s_iospec_bio3s_recv_R(t Place) (val IO_val)`. Since this function returns type `IO_val` and we can rule out `Unit` and `Internal` IO values, we always get either `Pkt2` or  `Unsupported`. I though we decided that if and only if the actual packet is a valid SCION packet we choose `Pkt2` otherwise `Unsupported`.

Reply:
I am not sure I get what you mean. What I wanted to convey is that, in the current design, when we return `Unsupported` it would be good to actually have a body in the function that returns it -- to that end, we would need to figure out the purpose of the parameters in the constructor of `Unsupported`. I think this would be preferable to not having a body.

Reply:
All I wanted to say is that we have to use `Unsupported` and cannot use `option`. Providing a body for `Unsupported` is probably tricky. I think the best would be to discuss this in detail in our next meeting.

Reply:
> All I wanted to say is that we have to use `Unsupported` and cannot use `option`

I understood that part. What is not clear to me is why it would be tricky to provide the body. It seems to me that the only thing we need to figure out is the role of the second parameter in the constructor of `Unsupported`. Nonetheless, for now, we can just add a todo (and open an issue on the issue tracker about this) and continue making progress

Reply:
As discussed, let's replace this `TypeParameter` by `Unit` and let's add a body here